pkg/util/testutil/flake/flake_test.go

```diff
@@ -0,0 +1,140 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.Datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package flake
+
+import (
+	"math/rand"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+type mockTesting struct {
+	*testing.T
+
+	mutex         sync.RWMutex
+	skipCallCount int
+	failCallCount int
+}
+
+func newMockTesting(t *testing.T) *mockTesting {
+	return &mockTesting{
+		T: t,
+	}
+}
+
+func (mt *mockTesting) Skip(_ ...any) {
+	func() {
+		mt.mutex.Lock()
+		defer mt.mutex.Unlock()
+		mt.skipCallCount++
+	}()
+	mt.SkipNow()
+}
+
+func (mt *mockTesting) Fail() {
+	mt.mutex.Lock()
+	defer mt.mutex.Unlock()
+	mt.failCallCount++
+}
+
+func (mt *mockTesting) SkipCount() int {
+	mt.mutex.RLock()
+	defer mt.mutex.RUnlock()
+	return mt.skipCallCount
+}
+
+func (mt *mockTesting) FailCount() int {
+	mt.mutex.RLock()
+	defer mt.mutex.RUnlock()
+	return mt.failCallCount
+}
+
+var trueValue = true
+var falseValue = false
+
+func TestFlake(t *testing.T) {
+	t.Run("allow flake test", func(t *testing.T) {
+		mt := newMockTesting(t)
+		// set allow-flake-failure flag to true
+		allowFlakeFailure = &trueValue
+		skipFlake = &falseValue
+		wrapAndRunFlakyTest(mt)
+		assert.True(t, mt.Skipped())
+		assert.False(t, mt.Failed())
+		assert.Greater(t, mt.SkipCount(), 1)
+		assert.Equal(t, 0, mt.FailCount())
+	})
+	t.Run("skip flake test", func(t *testing.T) {
+		mt := newMockTesting(t)
+		// set skip-flake flag to true
+		allowFlakeFailure = &falseValue
+		skipFlake = &trueValue
+		wrapAndRunFlakyTest(mt)
+		assert.Equal(t, 1, mt.SkipCount())
+		assert.Equal(t, 0, mt.FailCount())
+		assert.True(t, mt.Skipped())
+		assert.False(t, mt.Failed())
+	})
+	t.Run("skip flake test when both allow and skip are set", func(t *testing.T) {
+		mt := newMockTesting(t)
+		// set allow-flake-failure and skip-flake flag to true
+		allowFlakeFailure = &trueValue
+		skipFlake = &trueValue
+		wrapAndRunFlakyTest(mt)
+		assert.Equal(t, 1, mt.SkipCount())
+		assert.Equal(t, 0, mt.FailCount())
+		assert.True(t, mt.Skipped())
+		assert.False(t, mt.Failed())
+	})
+
+	t.Run("allow flaky sub tests", func(t *testing.T) {
+		mt := newMockTesting(t)
+		allowFlakeFailure = &trueValue
+		skipFlake = &falseValue
+		wrapAndRunSubFlakyTest(mt)
+		assert.False(t, mt.Failed())
+	})
+}
+
+func wrapAndRunFlakyTest(t *mockTesting) {
+	t.Helper()
+	wg := sync.WaitGroup{}
+	// testing.T.Skip() calls runtime.Goexit() which terminates the goroutine
+	// run the test in a separate goroutine to avoid terminating `TestFlake` test
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		ft := Wrap(t)
+		for i := 0; i < 100; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				coin := flipCoin()
+				assert.Equal(ft, "heads", coin)
+			}()
+		}
+	}()
+	wg.Wait()
```

Comment:
You don't need two layers of goroutine and the following code works.
```go
	ft := Wrap(t)
	for i := 0; i < 100; i++ {
		wg := sync.WaitGroup{}
		wg.Add(1)
		go func() {
			defer wg.Done()
			coin := flipCoin()
			assert.Equal(ft, "heads", coin)
		}()
		wg.Wait()
	}
```
By having a single goroutine at a time (Note the location of `wg.Wait()`) you can remove `mutex`.
What do you think?

Reply:
I need to have two layers as I need `ft := Wrap(t)` to happen in a goroutine. `t.Skip` calls `runtime.Goexit()` underneath, that terminates the goroutine where it is called. 

Since I use this same function with `skip` mode, if I call `ft := Wrap(t)` from the test runner thread, it will terminate the actual test, `TestFlake`. 

I know it's confusing, there is the comment at the beginning of the first call to a goroutine, but I should make it more explicit.

Note also this is also for meta testing purpose, actual tests will wrap on the main thread and stop the main test on flake

Reply:
Without the first goroutine level, on the test with `skipTest` true we call skip on the main thread and skip the actual test, not just the metatest